base/slicearray.jl

```diff
@@ -0,0 +1,223 @@
+    """
+    axes::AX
+end
+
+unitaxis(::AbstractArray) = Base.OneTo(1)
+
+function Slices(A::P, slicemap::SM, ax::AX) where {P,SM,AX}
+    N = length(ax)
+    S = Base._return_type(view, Tuple{P, map((a,l) -> l === (:) ? Colon : eltype(a), axes(A), slicemap)...})
+    Slices{P,SM,AX,S,N}(A, slicemap, ax)
+end
+
+_slice_check_dims(N) = nothing
+function _slice_check_dims(N, dim, dims...)
+    1 <= dim <= N || throw(DimensionMismatch("Invalid dimension $dim"))
+    dim in dims && throw(DimensionMismatch("Dimensions $dims are not unique"))
+    _slice_check_dims(N,dims...)
+end
+
+@inline function _eachslice(A::AbstractArray{T,N}, dims::NTuple{M,Integer}, drop::Bool) where {T,N,M}
+    _slice_check_dims(N,dims...)
+    if drop
+        # if N = 4, dims = (3,1) then
+        # axes = (axes(A,3), axes(A,1))
+        # slicemap = (2, :, 1, :)
+        ax = map(dim -> axes(A,dim), dims)
+        slicemap = ntuple(dim -> something(findfirst(isequal(dim), dims), (:)), N)
+        return Slices(A, slicemap, ax)
+    else
+        # if N = 4, dims = (3,1) then
+        # axes = (axes(A,1), OneTo(1), axes(A,3), OneTo(1))
+        # slicemap = (1, :, 3, :)
+        ax = ntuple(dim -> dim in dims ? axes(A,dim) : unitaxis(A), N)
+        slicemap = ntuple(dim -> dim in dims ? dim : (:), N)
+        return Slices(A, slicemap, ax)
+    end
+end
+@inline function _eachslice(A::AbstractArray, dim::Integer, drop::Bool)
+    _eachslice(A, (dim,), drop)
+end
+
+"""
+    eachslice(A::AbstractArray; dims, drop=true)
+
+Create a [`Slices`](@ref) object that is an array of slices over dimensions `dims` of `A`, returning
+views that select all the data from the other dimensions in `A`. `dims` can either by an
+integer or a tuple of integers.
+
+If `drop = true` (the default), the outer `Slices` will drop the inner dimensions, and
+the ordering of the dimensions will match those in `dims`. If `drop = false`, then the
+`Slices` will have the same dimensionality as the underlying array, with inner
+dimensions having size 1.
+
+See also [`eachrow`](@ref), [`eachcol`](@ref), and [`selectdim`](@ref).
+
+!!! compat "Julia 1.1"
+     This function requires at least Julia 1.1.
+
+!!! compat "Julia 1.7"
+     Prior to Julia 1.7, this returned an iterator, and only a single dimension `dims` was supported.
+
+# Example
+
+```jldoctest
+julia> M = [1 2 3; 4 5 6; 7 8 9]
+3×3 Matrix{Int64}:
+ 1  2  3
+ 4  5  6
+ 7  8  9
+
+julia> S = eachslice(M, dims=1)
+3-element Rows{Matrix{Int64}, Tuple{Base.OneTo{Int64}}, SubArray{Int64, 1, Matrix{Int64}, Tuple{Int64, Base.Slice{Base.OneTo{Int64}}}, true}}:
+ [1, 2, 3]
+ [4, 5, 6]
+ [7, 8, 9]
+
+julia> S[1]
+3-element view(::Matrix{Int64}, 1, :) with eltype Int64:
+ 1
+ 2
+ 3
+
+julia> T = eachslice(M, dims=1, drop=false)
+3×1 Slices{Matrix{Int64}, Tuple{Int64, Colon}, Tuple{Base.OneTo{Int64}, Base.OneTo{Int64}}, SubArray{Int64, 1, Matrix{Int64}, Tuple{Int64, Base.Slice{Base.OneTo{Int64}}}, true}, 2}:
+ [1, 2, 3]
+ [4, 5, 6]
+ [7, 8, 9]
```

Comment:
To explain what `dims` does, perhaps this should show a higher-dimensional example. They are messy to print but showing just the sizes ought to make it clear, something like this:
```suggestion
julia> sizes(x) = (element = size(first(x)), container = size(x));

julia> eachslice(M, dims=1, drop=false) |> sizes
(element = (4,), container = (3, 1))

julia> A = rand(5, 7, 11, 13);

julia> eachslice(A; dims=(2,4)) |> sizes
(element = (5, 11), container = (7, 13))

julia> eachslice(A; dims=(4,2)) |> sizes
(element = (5, 11), container = (13, 7))

julia> eachslice(A; dims=(4,2), drop=false) |> sizes
(element = (5, 11), container = (1, 7, 1, 13))
```
(where I've changed M to be 3×4 so that size 3 is unambiguous, maybe 2×3 would be better.)